cpp/src_prims/stats/histogram.h

```diff
@@ -67,8 +65,39 @@ void gmemHist(int* bins, IdxT nbins, const DataT* data, IdxT n, BinnerOp op,
               cudaStream_t stream) {
   int nblks = ceildiv<int>(n, TPB);
   CUDA_CHECK(cudaMemsetAsync(bins, 0, nbins * sizeof(int), stream));
-  gemmHistKernel<DataT, BinnerOp, IdxT><<<nblks, TPB, 0, stream>>>(bins, data,
-                                                                   n, op);
+  gmemHistKernel<DataT, BinnerOp, IdxT>
+    <<<nblks, TPB, 0, stream>>>(bins, data, n, op);
+  CUDA_CHECK(cudaGetLastError());
+}
+
+template <typename DataT, typename BinnerOp, typename IdxT>
+__global__ void smemHistKernel(int* bins, const DataT* data, IdxT n,
+                               BinnerOp binner) {
+  extern __shared__ int sbins[];
+  auto tid = threadIdx.x + IdxT(blockIdx.x) * blockDim.x;
+  for (auto i = tid; i < nbins; i += blockDim.x) {
+    sbins[i] = 0;
+  }
+  __syncthreads();
+  auto stride = IdxT(blockDim.x) * blockDim.x;
+  for (auto i = tid; i < n; i += stride) {
+    int binId = binner(data[i], i);
+    atomicAdd(sbins + binId, 1);
+  }
+  __syncthreads();
+  for (auto i = tid; i < nbins; i += blockDim.x) {
+    atomicAdd(bins + i, sbins[i]);
```

Comment:
Do you want to check if the sbins[i] is > 0?  Could save significant global atomics.

Reply:
Thanks. I missed this one. Fixed now.